### PR TITLE
allow repeatable fields to be sent with form on ajax requests

### PR DESCRIPTION
**Pull request by pxpm**
refs: #3218 

Problem: When using `include_all_form_fields` with ajax fields the repeatable values in the form would not be sent with the request because they only got their value parsed when the `submit` button in the main form is pressed.

Solution: Create an event handler in repeatable fields that we can fire when doing the ajax request and parse the values of the repeatable instances to be sent along with the main form fields.

### CODE REVIEW

**Comment by o15a3d4l11s2** [comment]
I can confirm that the change works properly for `select2_from_ajax field`.

**Comment by o15a3d4l11s2** [comment]
I can confirm that the fix works properly for `select2_from_ajax` field.

@pxpm, by my mistake I thought that the field did not work properly, but it was me... sorry.

**Comment by pxpm** [comment]
I noticed that I was registering 1 event per "repeatable" section, and that could lead to inconsistencies depending on crud size. 

I refactored the code a bit so we only register 1 event in the form, and we find the inputs we need to update in a for loop.

File: src/resources/views/crud/fields/repeatable.blade.php
Comment (promatik) [inline src/resources/views/crud/fields/repeatable.blade.php:128]: @pxpm what should we do with the element?
Should it go on the request too?

**Review by promatik — CHANGES_REQUESTED** [review]

File: src/resources/views/crud/fields/repeatable.blade.php
Comment (promatik) [inline src/resources/views/crud/fields/repeatable.blade.php:125]: @pxpm, when does it enter the second part of the condition?
I wasn't able to test that part.

File: src/resources/views/crud/fields/repeatable.blade.php
Comment (o15a3d4l11s2) [reply src/resources/views/crud/fields/repeatable.blade.php:128]: @promatik, select2_from_ajax normally includes the data in the request:
> only send the current field through AJAX

so I believe it should be included.

When
```php
'include_all_form_fields'  => true
```
I would expect that all the fields are sent.

Comment (pxpm) [reply src/resources/views/crud/fields/repeatable.blade.php:128]: @promatik 

Indeed, that is the answer to: #3440 

I just didn't pushed it here because I was planning to create a separate PR for that issue. 

This PR does what it states, "allows the repeatable fields to be sent with ajax requests" and the little extra I left as a `TODO`, that is we already have the "trigger" there.

File: src/resources/views/crud/fields/repeatable.blade.php
Comment (pxpm) [reply src/resources/views/crud/fields/repeatable.blade.php:125]: `backpack_field` is also the event namespace for `backpack_field.deleted`, so to go to the second condition you have to have a ckeditor that listens for the `backpack_field.deleted`.

Also enters the second condition everytime `backpack_field` exists in the `events` where it checks if we have any `parse_value` event already registed. 

In dummy example, where you have 5 or 6 `repeatable` fields, it will attempt to register 5,6 times the event, in the second attempt it will enter the condition and fail, so don't register the event that many times, only the first time.

Let me know,
Pedro

File: src/resources/views/crud/fields/repeatable.blade.php
Comment (promatik) [reply src/resources/views/crud/fields/repeatable.blade.php:128]: Ok sure, makes sense 😅

File: src/resources/views/crud/fields/relationship/fetch_or_create.blade.php
Comment (promatik) [inline src/resources/views/crud/fields/relationship/fetch_or_create.blade.php:275]: @pxpm after a huge debug, I found out `fetch_or_create` was not working properly because this code is inside `$inlineCreateButtonElement.on('click'` function, and it should be inside `element.select2({data:` function.

**I added it below.**
I'm not sure if this code is necessary here for anything related with inline create, can you make sure?

**Review by promatik — CHANGES_REQUESTED** [review]

File: src/resources/views/crud/fields/relationship/fetch_or_create.blade.php
Comment (pxpm) [reply src/resources/views/crud/fields/relationship/fetch_or_create.blade.php:275]: @promatik indeed, you are right, it should be inside the `element.select2` too. 

- One is the regular field when doing the search for something to select (`include_all_form_fields`) in the main field.
-  the other is when you make the request for inline create modal,
 ```php
'inline_create' => [
    'include_main_form_fields' => true
]
``` 
Nice catch!

**Comment by pxpm** [comment]
For me this is ready to merge! If you agree with me @promatik , just move it there.

Best,
Pedro

**Comment by promatik** [comment]
> For me this is ready to merge! If you agree with me @promatik , just move it there.
> 
> Best,
> Pedro

Tested once again, working as expected 👌

**Review by promatik — APPROVED** [review]

**Comment by aberaldo** [comment]
@tabacitu  

Hello! in my backpack project i overrided those 5 blade files (that where modified) in: 
resources/views/vendor/backpack/crud/fields/repeatable.blade.php
resources/views/vendor/backpack/crud/fields/select2_from_ajax_multiple.blade.php
resources/views/vendor/backpack/crud/fields/select2_from_ajax.blade.php
resources/views/vendor/backpack/crud/fields/relationship/fetch_or_create.blade.php
resources/views/vendor/backpack/crud/fields/relationship/fetch.blade.php

When the repeatable is loaded in the view i got:
image

The same when on edit and create.

Versions:
image

Maybe its requiring another fix or issue?

Thanks!

**Comment by pxpm** [comment]
@aberaldo could not reproduce your issue. 

Can you provide some sample code on how you setup your repeatable fields so I can try to reproduce ? 

Best,
Pedro

**Comment by aberaldo** [comment]
> @aberaldo could not reproduce your issue.
> 
> Can you provide some sample code on how you setup your repeatable fields so I can try to reproduce ?
> 
> Best,
> Pedro

Pedro, i didnt wrote any code. Only in my backpack project i override those 5 blade files.

**Comment by pxpm** [comment]
@aberaldo ok, but the error you reported can only happen in a repeatable field, so the `create/update` that is failing for you have a repeatable field ?

**Comment by aberaldo** [comment]
@pxpm yes. 
**Before:**
image

**After:**
image


The repeatable field is in the footer.
Thanks!

**Comment by pxpm** [comment]
Thanks for clarification @aberaldo , It really helped me to debug this issue. 

I cannot tell 100% what's the rootcause I am still trying to figure it out. 

If I add a repeatable field into `monsters` crud I have the error you described. If I go to `dummies`, (that only have repeatable fields) that error does not happen.

I will get back with my findings. :) 

Thanks again @aberaldo .

Best,
Pedro

**Comment by pxpm** [comment]
@aberaldo it's fixed. 

The problem is that if you have other fields that add events into `backpack_field` namespace when that code run, there is always some previous event. 

But in case you didn't use any of those fields the `form_events` would be `undefined` thus, triggering the error. 

Thanks again,
Pedro

**Comment by aberaldo** [comment]
@pxpm Yes its fixed!

Thanks!

**Comment by pxpm** [comment]
> @pxpm Yes its fixed!
> 
> Thanks!

I thank you for taking time to share the error and helping me to debug it. You are the one to thank :)

Best,
Pedro

**Review by tabacitu — CHANGES_REQUESTED** [review]
Sorry for delaying approving this, but something was triggering my spidey-sense and I couldn't quite put my finger on it. But now I think I know what was bothering me - what makes it feel a little off/hacky. 


**Quick 1)** I don't think `parse_value` is a good-enough name. 'cause it doesn't _just_ parse the value, right? It also updates the hidden inputs. That's what you _actually_ want from it, to calculate and update the value of the hidden inputs.

**Quirk 2)** The global variable/namespace `backpack_field`, which isn't used anywhere else.


-----

As a solution to both of the above:

What do you guys think about instead of `$form.trigger('backpack_field.parse_value', element);` we add the ability for ANY field to have a `data-update-function` that some other fields can call at will?

Let me rephrase it:
- we already have `data-init-function="bpFieldInitRepeatableElement"` (which isn't pretty at all, we probably should have used custom events here, but... it _is_ the way it works right now);
- we can add the ability for fields to similarly have `data-update-function="bpFieldUpdateRepeatableHiddenInput"` which would make sure to update the hidden inputs based on the visible inputs;
- not all fields would need to have this `data-update-function`, in fact MOST would not; but inside the fields/functions that need the final value, we could get all form children that have a `data-update-function` function and call those functions; 
- optionally, we could add a custom event listener on the form itself (say... `updateHiddenInputs`, to do that in one line where we need it, but... if that this event listener needs to be inside a JS file (not a Blade file) it'd be a breaking change so... maybe later;

----

Feel free to disagree with me - I'm not sure what I propose here is better. But it does seem more in line with how we've been doing things up until now.

**Comment by pxpm** [comment]
@tabacitu totally agree with the naming. It's not only `parse_value`, but it really is `parse_value_and_update_hidden_input`. 

About the `backpack_field` event namespace, it's already used in CKeditor: https://github.com/Laravel-Backpack/CRUD/blob/ecbe7676535d3c9121227329d62cad9978d29fee/src/resources/views/crud/fields/ckeditor.blade.php#L50

In this scenario it's used because of the deletion, so we might also need `data-on-delete-function` ? Using the events seemed easier, and also have the possibility to be used by any field that listen to this events. 

Also, I am not sure that using functions we can get the `element` that triggered that `on-update`. This is mainly used to filter the results of some select with a previous selection, in a repeatable you can have multiple selects of the same entity (1 per repeatable row).

Let me know,
Pedro

File: src/resources/views/crud/fields/repeatable.blade.php
Comment (pxpm) [inline src/resources/views/crud/fields/repeatable.blade.php:137]: @prodixx here, we need to pass this `element` row, so in the controller that receives the ajax request we know what constrain to apply to show the options.

Comment (o15a3d4l11s2) [reply src/resources/views/crud/fields/repeatable.blade.php:137]: @pxpm, I think you meant to mention @promatik? :)

Comment (pxpm) [reply src/resources/views/crud/fields/repeatable.blade.php:137]: Hello @o15a3d4l11s2 

Hope you are doing well. 
Actually this was intended to @prodixx . I was having a conversation with him on our gitter channel :)

You should join us there, not every day, but sometimes we have good chats there, and some improvements in backpack comes from user feedback there :)

Anyway 10 points to you from awarenesses. 👍 

Have a nice weekend,
Pedro

**Comment by tabacitu** [comment]
> About the backpack_field event namespace, it's already used in CKeditor:

> In this scenario it's used because of the deletion, so we might also need data-on-delete-function ? Using the events seemed easier, and also have the possibility to be used by any field that listen to this events.

Aaahhh 😡😅 Fine!

I may have been overthinking then, thanks for being patient @pxpm 🙏😅 Then I guess `backpack_field.parse_value` isn't that bad a name. Anyway, we can refactor in 4.2 or v12 if we find a better name.

If you and @promatik say this is ready and we should merge it, I'll merge it 👏👏👏

**Comment by promatik** [comment]
Tested once again, just to make sure 😅
@pxpm please give it a final approval ✌

**Comment by pxpm** [comment]
@promatik I think what is proposed here is totally fine and working, I'v just tested it. 

Why I didn't move this yet to ready to merge is because there is one bit here that I think it's missing before this becomes 100% ready. 

Is passing the element that triggered the form. 

This functionality is required for repeatable and it's only beeing implemented because of that, the thing is that this is only usefull if you know what element triggered the request so you can act accordingly. 

What I mean is that if you have 5 repetable items, this event is triggered, at the moment all the fields are passed in the request as it should, but you still don't know exactly which row triggered the request, so you can't filter the results by that same field.

I have a local branch where I am working on it. 

I will update this soon as I have a working solution for this. At the current state this will do the "heavy lifting" of sendiing the repeatable fields, but developer would be unable to know which row triggered the request. 

Best,
Pedro

**Comment by promatik** [comment]
@pxpm my bad, I thought you wanted to do that on a separate PR, merging this one before. But yeah, makes sense.
For the sake of testing please do it on a separate PR 😅

**Comment by o15a3d4l11s2** [comment]
@pxpm, is there any change that this can be finished in the near future? Is there any way I can support you so that it can be completed soon?

I have exactly the case that you describe - a `Product` model that has many feature values describing it. So I have a repeatable with two selects (value depends on the feature) like this:
| Feature  | Value |
| ------------- | ------------- |
| Color  | Red |
| Finish  | Matte |

The problem at the moment is that the `Value` select contains all possible data from the table, as I cannot determine **which** row the user is on, so I cannot filter and select only the data for the right `Feature->values`

**Comment by pxpm** [comment]
@tabacitu @promatik 

merged master here. 

This PR would allow repeatable fields to be sent along when querying ajax endpoints. It also introduces a new key `ajax_trigger_input` that is sent along with the request that contains the input name that triggered the request, and in case inside repeatable it also sends the row number where that input is located.

I've choosen a strange and long name on purpose so we avoid (at best we can) mess with other user input names.

Also sent https://github.com/Laravel-Backpack/docs/pull/260 to the DOCS to add this functionality there.

Please review and let me know,
Pedro

File: src/resources/views/crud/fields/select2_from_ajax.blade.php
Comment (promatik) [inline src/resources/views/crud/fields/select2_from_ajax.blade.php:173]: @pxpm when do we may have `data.data`?

Would it be better to have something like;
```js
if(data.data) {
  data = data.data;
}
// or
data = data.data || data;
```

That way we would avoid repeating the code, and we would be encapsulating that hack (not sure if it's some sort of hack or not).

Comment (pxpm) [reply src/resources/views/crud/fields/select2_from_ajax.blade.php]: I've just submitted a [commit](https://github.com/Laravel-Backpack/CRUD/pull/3439/commits/042ae6f627beaf33cdb1def3e81732c6092a463d) that addresses this. You are right at that. I am not sure it's hacky, it's just the way we separate the the paginated from non paginated results to define the `pagination: more` configuration later. 

Let me know what you think,
Pedro

File: src/resources/views/crud/fields/select2_from_ajax_multiple.blade.php
Comment (tabacitu) [inline src/resources/views/crud/fields/select2_from_ajax_multiple.blade.php:140]:
```suggestion
                                ajax_request_was_triggered_by: 
                                {
                                    'row': element.attr('data-row-number'), 
                                    'element': $fieldName
                                 }
```

@pxpm what do you think about renaming the variable like so? What bothered me was that `ajax_trigger_input` doesn't look like an array or object... you expect it to contain the name of the input, but then it returns something else.

Not sure this is any better. It certainly ain't prettier 😅

File: src/resources/views/crud/fields/repeatable.blade.php
Comment (tabacitu) [reply src/resources/views/crud/fields/repeatable.blade.php:168]: Ok soo... I've still got a problem with the name `backpack_field.parse_value`, sorry 😅👀😔

I propose we rename it to `backpack_field.calculate`. That way, we leave ourselves room to have more fields use it, as part of an entire cycle of supported events on fields:
- `backpack_field.init` / `backpack_field.initialize`, then `backpack_field.initalized`
- `backpack_field.calculate`, then `backpack_field.calculated`
- `backpack_field.hide`, then `backpack_field.hidden`
- `backpack_field.show`, then `backpack_field.shown`
- `backpack_field.delete`, then `backpack_field.deleted`

I wonder if we're going to remember this though... 😅 Or be able to find it here.

Alternatively... `backpack_field.compute`? Sorry for being a bitch about this @pxpm ... naming is hard... And I don't want us to have to create a breaking change for this in the next version, just because of the name.

**Review by tabacitu — CHANGES_REQUESTED** [review]
So sorry @pxpm ... I have a problem with the names again. Otherwise, the code is now clean in my opinion and easy to read. If @promatik also says it works after it's renamed, I'll merge it right away! Thank you for putting up with me... 😔

**Comment by tabacitu** [comment]
Update & note to selves: we've decided to scrap this one, in favour of rewriting `repeatable` to use array names in 4.2 (eg. `testimonial[0][name]`). If we do that, the only thing needed from this PR will be to send the row number and field name with AJAX.

Let's leave this PR open until that's merged though.

**Comment by tabacitu** [comment]
If anybody's using this branch directly, please let us know. We're not going to delete the branch any time soon, but we just want to know so we can give a heads-up before we do do that.

**Comment by tabacitu** [comment]
Can't wait to finally rewrite that bit of repeatable. Then validation for repeatable will be super-duper easy to do, instead of the clusterfuck it is now. Good idea @pxpm !

**Comment by tabacitu** [comment]
@pxpm I believe we can close this, because PR #3905 is going to it "just work", without any workarounds. Correct? 

If so, please close this.

**Comment by tabacitu** [comment]
Again, I think this is no longer needed due to #3905 being merged into 4.2. After that, subfields are real fields, with real names. So I'm pretty sure they'll get included in the AJAX request.

@pxpm if I'm wrong please re-open.

**Comment by CamusMX7** [comment]
@pxpm I've followed the topic from other posts and you mentioned "repeatable_row_trigger" and "ajax_trigger_input" in this case, but I understand they didn't apply.
Currently how can we get the row that triggered the request?
I mean the index of subfield, for example: $form['prices'][?]['category_price_id']

**Comment by pxpm** [comment]
Hello @CamusMX7 you can get it with `request('triggeredBy`)
I added it yesterday or something like to the docs. 

Sorry

**Comment by CamusMX7** [comment]
Really thank you very much for your help. You've already saved me a couple of times.
I implemented an alternative solution by creating a hidden field and sending it the value of the element that makes the request and thus obtain it in my FETCH function. But it didn't fully work...

I have a couple of possible minor bugs, but haven't had time to post them. One has to do with this issue of repeated subfields when the element has two "select2" and one depends on the other, the value of "aria-disabled" disables the value of the child with that same ID, that is if the parent select2 ( category) has selected ID 1, in the child select2 (article) the element with ID 1 is no longer available.
But I hope to confirm the bug and will publish it soon.

On the other hand, I am concerned about a recent case that was already published yesterday:
https://github.com/Laravel-Backpack/PermissionManager/issues/306

I am totally grateful for all your help and I hope that I will be lucky enough to see the publications that I make.

**Comment by pxpm** [comment]
Please, if you have any issues open a GitHub issue with your problem and we will try to figure out the best solution together. 
Creating an hidden field for this might have undesired consequences and/or force you to configure the crud to respect your changes. I'd recommend using the "official" crud stuff when available unless is something so custom that we can't incorporate. 
Glad I can help, let me know any issue you are having :)
Cheers

**Pull request closed without merging**
